compiler/ast/dag/keyfilter.go

```diff
@@ -0,0 +1,119 @@
+package dag
+
+import (
+	"github.com/brimdata/zed/order"
+	"github.com/brimdata/zed/pkg/field"
+)
+
+type KeyFilter struct {
+	Expr Expr
+}
+
+// NewKeyFilter creates a KeyFilter that contains an modified form of node where
+// only predicates operating on key are kept. The underlying expression is
+// rewritten in a manner so that results produced by the filter will always be
+// a superset of the results produced by the parent filter (i.e., it will not
+// filter values that would not be not also filtered by the original filter).
+// Currently KeyFilter only recognizes simple key predicates against a literal
+// value and using the comparators ==, >=, >, <, <=, otherwise the predicate is
+// ignored.
+func NewKeyFilter(key field.Path, node Expr) *KeyFilter {
+	e := visitLeaves(node, func(cmp string, lhs *This, rhs *Literal) Expr {
+		if !key.Equal(lhs.Path) {
+			return nil
+		}
+		return NewBinaryExpr(cmp, lhs, rhs)
+	})
+	if e == nil {
+		return nil
+	}
+	return &KeyFilter{e}
+}
+
+// SpanFilter creates an Expr that returns true if the KeyFilter has a value
+// within a span of values. The span compared against must be a record with the
+// fields "lower" and "upper", where "lower" is the inclusive lower bounder and
+// "upper" is the exclusive upper bound.
+func (k *KeyFilter) SpanFilter(o order.Which, prefix ...string) Expr {
+	lower := append([]string{}, append(prefix, "lower")...)
+	upper := append([]string{}, append(prefix, "upper")...)
```

Comment:
Copy `prefix` before appending to avoid unexpectedly modifying its underlying array.
```suggestion
	lower := append(append([]string{}, prefix...), "lower")
	upper := append(append([]string{}, prefix...), "upper")
```